src/qibolab/transpilers/general_connectivity.py

```diff
@@ -0,0 +1,393 @@
+
+import matplotlib.pyplot as plt
+import networkx as nx
+import numpy as np
+from more_itertools import pairwise
+from qibo import gates
+from qibo.config import log, raise_error
+from qibo.models import Circuit
+
+DEFAULT_INIT_SAMPLES = 100
+
+
+class QubitInitMethod(Enum):
+    """A class to define the initial qubit mapping methods."""
+
+    greedy = auto()
+    subgraph = auto()
+    custom = auto()
+
+
+class Transpiler:
+    """A class to perform initial qubit mapping and connectivity matching.
+
+    Properties:
+        connectivity (networkx.graph): chip connectivity.
+        init_method (string or QubitInitMethod): initial qubit mapping method.
+        init_samples (int): number of random qubit initializations for greedy initial qubit mapping.
+
+    Attributes:
+        _circuit_repr (list): quantum circuit represented as a list (only 2 qubit gates).
+        _mapping (dict): circuit to physical qubit mapping during transpiling.
+        _graph (networkx.graph): qubit mapped as nodes of the connectivity graph.
+        _qubit_map (np.array): circuit to physical qubit mapping during transpiling as vector.
+        _circuit_position (int): position in the circuit.
+        _added_swaps (int): number of swaps added to the circuit to match connectivity.
+    """
+
+    def __init__(self, connectivity, init_method="greedy", init_samples=None):
+        self.connectivity = connectivity
+        self.init_method = init_method
+        if self.init_method is QubitInitMethod.greedy and init_samples is None:
+            init_samples = DEFAULT_INIT_SAMPLES
+        self.init_samples = init_samples
+
+        self._circuit_repr = None
+        self._mapping = None
+        self._graph = None
+        self._qubit_map = None
+        self._circuit_position = 0
+        self._added_swaps = 0
+
+    def transpile(self, qibo_circuit):
+        """Qubit mapping initialization and circuit transpiling.
+
+        Args:
+            qibo_circuit (qibo.Circuit): circuit to be transpiled.
+
+        Returns:
+            hardware_mapped_circuit (qibo.Circuit): circut mapped to hardware topology.
+            final_mapping (dict): final qubit mapping.
+            init_mapping (dict): initial qubit mapping.
+            added_swaps (int): number of swap gates added.
+        """
+        self._circuit_position = 0
+        self._added_swaps = 0
+        self.translate_circuit(qibo_circuit)
+        keys = list(self._connectivity.nodes())
+        if self._init_method is QubitInitMethod.greedy:
+            self.greedy_init()
+        elif self._init_method is QubitInitMethod.subgraph:
+            self.subgraph_init()
+        elif self._init_method is QubitInitMethod.custom:
+            self._mapping = dict(zip(keys, self._mapping.values()))
+            self._graph = nx.relabel_nodes(self._connectivity, self._mapping)
+        # Inverse permutation
+        init_qubit_map = np.argsort(np.asarray(list(self._mapping.values())))
```

Comment:
Here you have two options to slightly simplify the stack of calls:
1. you are using `list()` because `np.asarray()` does not recognize `.values()` as a sequence, but then you can use `np.fromiter()` (that instead requires a specified `dtype`) 
```suggestion
        init_qubit_map = np.argsort(np.fromiter(self._mapping.values(), int))
```
2. just skip `np.asarray()`, since `np.argsort()` accepts an `array_like`, and a list is included (moreover, the function returns a new array, and it does not modify the input one)
```suggestion
        init_qubit_map = np.argsort(list(self._mapping.values()))
```

Reply:
This is just simplifying the code, and the speed improvement is marginal.

It would be more relevant as memory consumption improvement, but it is only significant for large arrays, since at the moment you are constructing three copies of the input values:
1. `.values()`, the original (that is most likely a view on the `dict` values, so this call does not consume more memory)
2. `list()`, that construct a list out of the elements, so it allocates memory for a copy
3. `np.asarray()`, that takes the list and copy in a new array.

However, for small arrays, the overhead is negligible. Just good to know it is happening.